kyotocabinet/man/kcutilmgr.1
.TH "KCUTILMGR" 1 "2011-03-04" "Man Page" "Kyoto Cabinet"

.SH NAME
kcutilmgr \- command line interface of miscellaneous utilities

.SH DESCRIPTION
.PP
.PP
The command `\fBkcutilmgr\fR' is a tool of miscellaneous utilities, and to show the configuration.  This command is used in the following format.  `\fIfile\fR' specifies a input file.  If it is omitted, the standard input is read.  If it begins with "@", the trailing substring is treated as the input.  `\fIpattern\fR' specifies an matching pattern.
.PP
.RS
.br
\fBkcutilmgr hex \fR[\fB\-d\fR]\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Performs hexadecimal encoding and its decoding.
.RE
.br
\fBkcutilmgr enc \fR[\fB\-hex\fR|\fB\-url\fR|\fB\-quote\fR]\fB \fR[\fB\-d\fR]\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Performs Base64 encoding and its decoding.
.RE
.br
\fBkcutilmgr ciph \fR[\fB\-key \fIstr\fB\fR]\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Performs Arcfour cipher and its decipher.
.RE
.br
\fBkcutilmgr comp \fR[\fB\-def\fR|\fB\-gz\fR|\fB\-lzo\fR|\fB\-lzma\fR]\fB \fR[\fB\-d\fR]\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Performs ZLIB encoding and its decoding.  By default, use the raw format.
.RE
.br
\fBkcutilmgr hash \fR[\fB\-fnv\fR|\fB\-path\fR|\fB\-crc\fR]\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Calculates the hash value.  By default, use MurMur hashing.
.RE
.br
\fBkcutilmgr regex \fR[\fB\-alt \fIstr\fB\fR]\fB \fR[\fB\-ic\fR]\fB \fIpattern\fB \fR[\fB\fIfile\fB\fR]\fB\fR
.RS
Prints lines matching a regular expression.
.RE
.br
\fBkcutilmgr conf \fR[\fB\-v\fR|\fB\-i\fR|\fB\-l\fR|\fB\-p\fR]\fB\fR
.RS
Shows the configuration of Kyoto Cabinet.
.RE
.br
\fBkcutilmgr version\fR
.RS
Shows the version information of Kyoto Cabinet.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-d\fR : perform decoding (unescaping), not encoding (escaping).
.br
\fB\-hex\fR : use hexadecimal encoding.
.br
\fB\-url\fR : use URL encoding.
.br
\fB\-quote\fR : use Quoted\-printable encoding.
.br
\fB\-key \fIstr\fR\fR : set the cipher key.
.br
\fB\-def\fR : use the deflate format.
.br
\fB\-gz\fR : use the gzip format.
.br
\fB\-lzo\fR : use LZO encoding.
.br
\fB\-lzma\fR : use LZMA encoding.
.br
\fB\-fnv\fR : use FNV hashing.
.br
\fB\-path\fR : use the path hashing of the directory database.
.br
\fB\-crc\fR : calculate the CRC32 checksum.
.br
\fB\-alt \fIstr\fR\fR : replaces matching substring with the alternative string.
.br
\fB\-ic\fR : ignores difference between upper and lower cases.
.br
\fB\-v\fR : show the version number of Kyoto Cabinet.
.br
\fB\-i\fR : show options to include the headers of Tokyo Cabinet.
.br
\fB\-l\fR : show options to link the library of Tokyo Cabinet.
.br
\fB\-p\fR : show the directory path of the commands.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR kcutiltest (1)


kyotocabinet/ChangeLog
2011-03-01  FAL Labs  <info@fallabs.com>

	* kclangc.h (kcmapnew, kcmapdel, kcmapiterator, kcmapsorter): new functions.

	* kclangctest.cc (runmap, procmap): new functions.

	- Release: 1.2.48

2011-02-25  FAL Labs  <info@fallabs.com>

	* kcdb.h (BasicDB::occupy): new function.

	- Release: 1.2.47

2011-02-23  FAL Labs  <info@fallabs.com>

	* kcutil.h (SIZEMAX, FLTMAX, DBLMAX): new constants.

	* kcplantdb.h (PlantDB::open): repair mechanism was modified.

	* kcplantdb.h (PlantDB::Cursor::accept_spec): a bug related to transactin was fixed.

	- all database classes were refactored to support some old compilers.

	- Release: 1.2.46

2011-02-21  FAL Labs  <info@fallabs.com>

	* kcdbext.h (MapReduce::execute): adaptive comparator was added.

	* kcdbext.h (MapReduce::execute_reduce): new function.

	* kcdbext.h (MapReduce::before, MapReduce::after): new functions.

	- Release: 1.2.45

2011-02-15  FAL Labs  <info@fallabs.com>

	* kcdb.h (DB::Visitor::visit_before, DB::Visitor::visit_after): new functions.

	- configuration files were modified for affinity for binary distributions.

	- all classes were refactored to abolish unnamed namespaces.

	- Release: 1.2.44

2011-02-12  FAL Labs  <info@fallabs.com>

	* kchashdb.h (HashDB): the type of record locking objects was changed.

	* kcdirdb.h (DirDB): the type of record locking objects was changed.

	* kcplantdb.h (PlantDB): the type of page locking objects was changed.

	- Release: 1.2.43

2011-02-10  FAL Labs  <info@fallabs.com>

	* kcdb.h (DB::Cursor::get_key, DB::Cursor::get_value): signature was modified.

	* kcdb.h (DB::Cursor::get_pair, DB::Cursor::get): renamed.

	* kcdb.h (DB::get): signature was modified.

	- Release: 1.2.42

2011-02-07  FAL Labs  <info@fallabs.com>

	* kccommon.h: macros for integer range were added.

	* kchashdb.h (HashDB::begin_transaction): behavior of locking was modified.

	* kcdirdb.h (DirDB::begin_transaction): behavior of locking was modified.

	- Release: 1.2.41

2011-02-05  FAL Labs  <info@fallabs.com>

	* kccompare.h (LexicalDescendingComparator, LexicalDescendingComparator): new classes.

	* kccompare.h (LEXICALCOMP, DECIMALCOMP): the type was changed to pointer.

	* kccompress.h (ZLIBRAWCOMP): the type was changed to pointer.

	* kcplantdb.h (PlantDB::load_meta): a bug related to custom comparator was fixed.

	- Release: 1.2.40

2011-02-01  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::open): checking type type was added.

	* kcplantdb.h (Plantdb::reorganize_file): recovery condition was modified.

	* kcpolymgr.cc (proccheck): a bug related to path handling was fixed.

	- Release: 1.2.39

2011-01-30  FAL Labs  <info@fallabs.com>

	* kcthread.cc: threshold of sleep locking was modified.

	* kcplantdb.h (PlantDB::accept): a bug related to race condition was fixed.

	* kcplantdb.h (PlantDB::save_leaf_node): locking mode was modified.

	- Release: 1.2.38

2011-01-25  FAL Labs  <info@fallabs.com>

	* kcutil.h (strvecdump, strvecload, strmapdump, strmapload): new functions.

	- Release: 1.2.37

2011-01-21  FAL Labs  <info@fallabs.com>

	* kcutil.cc (getsysinfo): a bug related to memory stats was fixed.

	* kcthread.cc (Thread::chill): new function.

	* kcthread.cc (ScopedMutex, ScopedSpinLock): refactored using primitives.

	* kcthread.cc (ScopedRWLock, ScopedSpinRWLock): refactored using primitives.

	- Release: 1.2.36

2011-01-11  FAL Labs  <info@fallabs.com>

	* kcutil.cc: the order of header inclusion was modified.

	* kcthread.cc (SpinRWLock): space efficiency was improved.

	- Release: 1.2.35

2011-01-09  FAL Labs  <info@fallabs.com>

	* kcplantdb.h (PlantDB::accept, PlantDB::accept_bulk): behavior of async was modified.

	- Release: 1.2.34

2011-01-05  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::open): handling the WAL file was modified.

	* kchashdb.h (HashDB::clear): handling the open flag was added.

	* kcplantdb.h (PlantDB::clear): handling the recovery flag was modified.

	- Release: 1.2.33

2011-01-03  FAL Labs  <info@fallabs.com>

	* kcfile.cc (File::read_fast): error check was added.

	* kclangc.cc (kcdbacceptbulk, kcdbsetbulk, kcdbremovebulk, kcdbgetbulk): new functions.

	- Release: 1.2.32

2010-12-30  FAL Labs  <info@fallabs.com>

	* kcdb.h (BasicDB::accept_bulk): new function.

	* kcdb.h (BasicDB::set_bulk, BasicDB::remove_bulk, BasicDB::get_bulk): new functions.

	* kcthread.h (SlottedMutex::unlock_all): the order became forward.

	* kcpolymgr.cc (procsetbulk, procremovebulk, procgetbulk): new functions.

	- Release: 1.2.31

2010-12-19  FAL Labs  <info@fallabs.com>

	* kcpolydb.h (PolyDB::set_internal_db): new function.

	- Release: 1.2.30

2010-12-11  FAL Labs  <info@fallabs.com>

	* kccommon.h (modfl): new function for portability to cygwin.

	* kcthread.cc (Mutex::lock_try): cygwin was added to work around cases.

	* cmdcommon.h (oprintf): new function instead of "iprintf".

	- Release: 1.2.29

2010-12-10  FAL Labs  <info@fallabs.com>

	* ktutil.h (atoin, atofn): new functions.

	- Release: 1.2.28


kyotocabinet/kcutilmgr.cc
/*************************************************************************************************
 * The command line interface of miscellaneous utilities
 *                                                               Copyright (C) 2009-2011 FAL Labs
 * This file is part of Kyoto Cabinet.
 * This program is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version
 * 3 of the License, or any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *************************************************************************************************/


#include "cmdcommon.h"


// global variables
const char* g_progname;                  // program name


// function prototypes
int main(int argc, char** argv);
static void usage();
static int32_t runhex(int argc, char** argv);
static int32_t runenc(int argc, char** argv);
static int32_t runciph(int argc, char** argv);
static int32_t runcomp(int argc, char** argv);
static int32_t runhash(int argc, char** argv);
static int32_t runregex(int argc, char** argv);
static int32_t runconf(int argc, char** argv);
static int32_t prochex(const char* file, bool dec);
static int32_t procenc(const char* file, int32_t mode, bool dec);
static int32_t procciph(const char* file, const char* key);
static int32_t proccomp(const char* file, int32_t mode, bool dec);
static int32_t prochash(const char* file, int32_t mode);
static int32_t procregex(const char* file, const char* pattern, const char* alt, int32_t opts);
static int32_t procconf(int32_t mode);


// main routine
int main(int argc, char** argv) {
  g_progname = argv[0];
  kc::setstdiobin();
  if (argc < 2) usage();
  int32_t rv = 0;
  if (!std::strcmp(argv[1], "hex")) {
    rv = runhex(argc, argv);
  } else if (!std::strcmp(argv[1], "enc")) {
    rv = runenc(argc, argv);
  } else if (!std::strcmp(argv[1], "ciph")) {
    rv = runciph(argc, argv);
  } else if (!std::strcmp(argv[1], "comp")) {
    rv = runcomp(argc, argv);
  } else if (!std::strcmp(argv[1], "hash")) {
    rv = runhash(argc, argv);
  } else if (!std::strcmp(argv[1], "regex")) {
    rv = runregex(argc, argv);
  } else if (!std::strcmp(argv[1], "conf")) {
    rv = runconf(argc, argv);
  } else if (!std::strcmp(argv[1], "version") || !std::strcmp(argv[1], "--version")) {
    printversion();
    rv = 0;
  } else {
    usage();
  }
  return rv;
}


// print the usage and exit
static void usage() {
  eprintf("%s: command line interface of miscellaneous utilities of Kyoto Cabinet\n",
          g_progname);
  eprintf("\n");
  eprintf("usage:\n");
  eprintf("  %s hex [-d] [file]\n", g_progname);
  eprintf("  %s enc [-hex|-url|-quote] [-d] [file]\n", g_progname);
  eprintf("  %s ciph [-key str] [file]\n", g_progname);
  eprintf("  %s comp [-def|-gz|-lzo|-lzma] [-d] [file]\n", g_progname);
  eprintf("  %s hash [-fnv|-path|-crc] [file]\n", g_progname);
  eprintf("  %s regex [-alt str] [-ic] pattern [file]\n", g_progname);
  eprintf("  %s conf [-v|-i|-l|-p]\n", g_progname);
  eprintf("  %s version\n", g_progname);
  eprintf("\n");
  std::exit(1);
}


// parse arguments of hex command
static int32_t runhex(int argc, char** argv) {
  bool argbrk = false;
  const char* file = NULL;
  bool dec = false;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-d")) {
        dec = true;
      } else {
        usage();
      }
    } else if (!file) {
      argbrk = true;
      file = argv[i];
    } else {
      usage();
    }
  }
  int32_t rv = prochex(file, dec);
  return rv;
}


// parse arguments of enc command
static int32_t runenc(int argc, char** argv) {
  bool argbrk = false;
  const char* file = NULL;
  int32_t mode = 0;
  bool dec = false;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-hex")) {
        mode = 1;
      } else if (!std::strcmp(argv[i], "-url")) {
        mode = 2;
      } else if (!std::strcmp(argv[i], "-quote")) {
        mode = 3;
      } else if (!std::strcmp(argv[i], "-d")) {
        dec = true;
      } else {
        usage();
      }
    } else if (!file) {
      argbrk = true;
      file = argv[i];
    } else {
      usage();
    }
  }
  int32_t rv = procenc(file, mode, dec);
  return rv;
}


// parse arguments of ciph command
static int32_t runciph(int argc, char** argv) {
  bool argbrk = false;
  const char* file = NULL;
  const char* key = "";
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-key")) {
        if (++i >= argc) usage();
        key = argv[i];
      } else {
        usage();
      }
    } else if (!file) {
      argbrk = true;
      file = argv[i];
    } else {
      usage();
    }
  }
  int32_t rv = procciph(file, key);
  return rv;
}


// parse arguments of comp command
static int32_t runcomp(int argc, char** argv) {
  bool argbrk = false;
  const char* file = NULL;
  int32_t mode = 0;
  bool dec = false;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-def")) {
        mode = 1;
      } else if (!std::strcmp(argv[i], "-gz")) {
        mode = 2;
      } else if (!std::strcmp(argv[i], "-lzo")) {
        mode = 3;
      } else if (!std::strcmp(argv[i], "-lzma")) {
        mode = 4;
      } else if (!std::strcmp(argv[i], "-d")) {
        dec = true;
      } else {
        usage();
      }
    } else if (!file) {
      argbrk = true;
      file = argv[i];
    } else {
      usage();
    }
  }
  int32_t rv = proccomp(file, mode, dec);
  return rv;
}


// parse arguments of hash command
static int32_t runhash(int argc, char** argv) {
  bool argbrk = false;
  const char* file = NULL;
  int32_t mode = 0;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-fnv")) {
        mode = 1;
      } else if (!std::strcmp(argv[i], "-path")) {
        mode = 2;
      } else if (!std::strcmp(argv[i], "-crc")) {
        mode = 3;
      } else {
        usage();
      }
    } else if (!file) {
      argbrk = true;
      file = argv[i];
    } else {
      usage();
    }
  }
  int32_t rv = prochash(file, mode);
  return rv;
}


// parse arguments of regex command
static int32_t runregex(int argc, char** argv) {
  bool argbrk = false;
  const char* pattern = NULL;
  const char* file = NULL;
  const char* alt = NULL;
  int32_t opts = 0;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-alt")) {
        if (++i >= argc) usage();
        alt = argv[i];
      } else if (!std::strcmp(argv[i], "-ic")) {
        opts |= kc::Regex::IGNCASE;
      } else {
        usage();
      }
    } else if (!pattern) {
      argbrk = true;
      pattern = argv[i];
    } else if (!file) {
      file = argv[i];
    } else {
      usage();
    }
  }
  if (!pattern) usage();
  int32_t rv = procregex(file, pattern, alt, opts);
  return rv;
}


// parse arguments of conf command
static int32_t runconf(int argc, char** argv) {
  bool argbrk = false;
  int32_t mode = 0;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-v")) {
        mode = 'v';
      } else if (!std::strcmp(argv[i], "-i")) {
        mode = 'i';
      } else if (!std::strcmp(argv[i], "-l")) {
        mode = 'l';
      } else if (!std::strcmp(argv[i], "-p")) {
        mode = 'p';
      } else {
        usage();
      }
    } else {
      argbrk = true;
      usage();
    }
  }
  int32_t rv = procconf(mode);
  return rv;
}


// perform hex command
static int32_t prochex(const char* file, bool dec) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  if (dec) {
    char c;
    while (is->get(c)) {
      int32_t cc = (unsigned char)c;
      int32_t num = -1;
      if (cc >= '0' && cc <= '9') {
        num = cc - '0';
      } else if (cc >= 'a' && cc <= 'f') {
        num = cc - 'a' + 10;
      } else if (cc >= 'A' && cc <= 'F') {
        num = cc - 'A' + 10;
      }
      if (num >= 0) {
        if (is->get(c)) {
          cc = (unsigned char)c;
          if (cc >= '0' && cc <= '9') {
            num = num * 0x10 + cc - '0';
          } else if (cc >= 'a' && cc <= 'f') {
            num = num * 0x10 + cc - 'a' + 10;
          } else if (cc >= 'A' && cc <= 'F') {
            num = num * 0x10 + cc - 'A' + 10;
          }
          std::cout << (char)num;
        } else {
          std::cout << (char)num;
          break;
        }
      }
    }
    if (istr) std::cout << std::endl;
  } else {
    bool mid = false;
    char c;
    while (is->get(c)) {
      if (mid) std::cout << ' ';
      int32_t cc = (unsigned char)c;
      int32_t num = (cc >> 4);
      if (num < 10) {
        std::cout << (char)('0' + num);
      } else {
        std::cout << (char)('a' + num - 10);
      }
      num = (cc & 0x0f);
      if (num < 10) {
        std::cout << (char)('0' + num);
      } else {
        std::cout << (char)('a' + num - 10);
      }
      mid = true;
    }
    std::cout << std::endl;
  }
  return 0;
}


// perform enc command
static int32_t procenc(const char* file, int32_t mode, bool dec) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  bool err = false;
  switch (mode) {
    default: {
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::basedecode(ostr.c_str(), &zsiz);
        std::cout.write(zbuf, zsiz);
        delete[] zbuf;
        if (istr) std::cout << std::endl;
      } else {
        char* zbuf = kc::baseencode(ostr.data(), ostr.size());
        std::cout << zbuf;
        delete[] zbuf;
        std::cout << std::endl;
      }
      break;
    }
    case 1: {
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::hexdecode(ostr.c_str(), &zsiz);
        std::cout.write(zbuf, zsiz);
        delete[] zbuf;
        if (istr) std::cout << std::endl;
      } else {
        char* zbuf = kc::hexencode(ostr.data(), ostr.size());
        std::cout << zbuf;
        delete[] zbuf;
        std::cout << std::endl;
      }
      break;
    }
    case 2: {
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::urldecode(ostr.c_str(), &zsiz);
        std::cout.write(zbuf, zsiz);
        delete[] zbuf;
        if (istr) std::cout << std::endl;
      } else {
        char* zbuf = kc::urlencode(ostr.data(), ostr.size());
        std::cout << zbuf;
        delete[] zbuf;
        std::cout << std::endl;
      }
      break;
    }
    case 3: {
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::quotedecode(ostr.c_str(), &zsiz);
        std::cout.write(zbuf, zsiz);
        delete[] zbuf;
        if (istr) std::cout << std::endl;
      } else {
        char* zbuf = kc::quoteencode(ostr.data(), ostr.size());
        std::cout << zbuf;
        delete[] zbuf;
        std::cout << std::endl;
      }
      break;
    }
  }
  return err ? 1 : 0;
}


// perform ciph command
static int32_t procciph(const char* file, const char* key) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  char* cbuf = new char[ostr.size()];
  kc::arccipher(ostr.data(), ostr.size(), key, std::strlen(key), cbuf);
  std::cout.write(cbuf, ostr.size());
  delete[] cbuf;
  return 0;
}


// perform comp command
static int32_t proccomp(const char* file, int32_t mode, bool dec) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  bool err = false;
  switch (mode) {
    default: {
      kc::ZLIB::Mode zmode;
      switch (mode) {
        default: zmode = kc::ZLIB::RAW; break;
        case 1: zmode = kc::ZLIB::DEFLATE; break;
        case 2: zmode = kc::ZLIB::GZIP; break;
      }
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::ZLIB::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::ZLIB::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
    case 3: {
      kc::LZO::Mode zmode = kc::LZO::RAW;
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::LZO::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::LZO::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
    case 4: {
      kc::LZMA::Mode zmode = kc::LZMA::RAW;
      if (dec) {
        size_t zsiz;
        char* zbuf = kc::LZMA::decompress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: decompression failed\n", g_progname);
          err = true;
        }
      } else {
        size_t zsiz;
        char* zbuf = kc::LZMA::compress(ostr.data(), ostr.size(), &zsiz, zmode);
        if (zbuf) {
          std::cout.write(zbuf, zsiz);
          delete[] zbuf;
        } else {
          eprintf("%s: compression failed\n", g_progname);
          err = true;
        }
      }
      break;
    }
  }
  return err ? 1 : 0;
}


// perform hash command
static int32_t prochash(const char* file, int32_t mode) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  std::ostringstream oss;
  char c;
  while (is->get(c)) {
    oss.put(c);
  }
  const std::string& ostr = oss.str();
  switch (mode) {
    default: {
      uint64_t hash = kc::hashmurmur(ostr.data(), ostr.size());
      oprintf("%016llx\n", (unsigned long long)hash);
      break;
    }
    case 1: {
      uint64_t hash = kc::hashfnv(ostr.data(), ostr.size());
      oprintf("%016llx\n", (unsigned long long)hash);
      break;
    }
    case 2: {
      char name[kc::NUMBUFSIZ];
      uint32_t hash = kc::hashpath(ostr.data(), ostr.size(), name);
      oprintf("%s\t%08lx\n", name, (unsigned long)hash);
      break;
    }
    case 3: {
      uint32_t hash = kc::ZLIB::calculate_crc(ostr.data(), ostr.size());
      oprintf("%08x\n", (unsigned)hash);
      break;
    }
  }
  return 0;
}


// perform regex command
static int32_t procregex(const char* file, const char* pattern, const char* alt, int32_t opts) {
  const char* istr = file && *file == '@' ? file + 1 : NULL;
  std::istream *is;
  std::ifstream ifs;
  std::istringstream iss(istr ? istr : "");
  if (file) {
    if (istr) {
      is = &iss;
    } else {
      ifs.open(file, std::ios_base::in | std::ios_base::binary);
      if (!ifs) {
        eprintf("%s: %s: open error\n", g_progname, file);
        return 1;
      }
      is = &ifs;
    }
  } else {
    is = &std::cin;
  }
  if (alt) {
    kc::Regex regex;
    if (!regex.compile(pattern, opts)) {
      eprintf("%s: %s: compilation failed\n", g_progname, pattern);
      return 1;
    }
    std::string altstr(alt);
    std::string line;
    while (mygetline(is, &line)) {

      std::cout << regex.replace(line, altstr) << std::endl;
    }



  } else {
    kc::Regex regex;
    if (!regex.compile(pattern, opts | kc::Regex::MATCHONLY)) {
      eprintf("%s: %s: compilation failed\n", g_progname, pattern);
      return 1;
    }
    std::string line;
    while (mygetline(is, &line)) {
      if (regex.match(line)) std::cout << line << std::endl;
    }
  }
  return 0;
}


// perform conf command
static int32_t procconf(int32_t mode) {
  switch (mode) {
    case 'v': {
      oprintf("%s\n", kc::VERSION);
      break;
    }
    case 'i': {
      oprintf("%s\n", _KC_APPINC);
      break;
    }
    case 'l': {
      oprintf("%s\n", _KC_APPLIBS);
      break;
    }
    case 'p': {
      oprintf("%s\n", _KC_BINDIR);
      break;
    }
    default: {
      oprintf("VERSION: %s\n", kc::VERSION);
      oprintf("LIBVER: %d\n", kc::LIBVER);
      oprintf("LIBREV: %d\n", kc::LIBREV);
      oprintf("FMTVER: %d\n", kc::FMTVER);
      oprintf("SYSNAME: %s\n", kc::SYSNAME);
      oprintf("BIGEND: %d\n", kc::BIGEND);
      oprintf("CLOCKTICK: %d\n", kc::CLOCKTICK);
      oprintf("PAGESIZE: %d\n", kc::PAGESIZE);
      oprintf("FEATURES: %s\n", kc::FEATURES);
      oprintf("TYPES: void*=%d short=%d int=%d long=%d long_long=%d size_t=%d"
              " float=%d double=%d long_double=%d\n",
              (int)sizeof(void*), (int)sizeof(short), (int)sizeof(int), (int)sizeof(long),
              (int)sizeof(long long), (int)sizeof(size_t),
              (int)sizeof(float), (int)sizeof(double), (int)sizeof(long double));
      std::map<std::string, std::string> info;
      kc::getsysinfo(&info);
      if (info["mem_total"].size() > 0)
        oprintf("MEMORY: total=%s free=%s cached=%s\n",
                info["mem_total"].c_str(), info["mem_free"].c_str(),
                info["mem_cached"].c_str());
      if (std::strcmp(_KC_PREFIX, "*")) {
        oprintf("prefix: %s\n", _KC_PREFIX);
        oprintf("includedir: %s\n", _KC_INCLUDEDIR);
        oprintf("libdir: %s\n", _KC_LIBDIR);
        oprintf("bindir: %s\n", _KC_BINDIR);
        oprintf("libexecdir: %s\n", _KC_LIBEXECDIR);
        oprintf("appinc: %s\n", _KC_APPINC);
        oprintf("applibs: %s\n", _KC_APPLIBS);
      }
      break;
    }
  }
  return 0;
}



// END OF FILE
